src/gui/inputdevices/InputUtils.h

```diff
@@ -1,9 +1,43 @@
-#include "control/ToolHandler.h"
-#include "control/settings/Settings.h"
+/**
+ * @file InputUtils.h
+ * @author idotobi
+ * @brief Helper functions in use in the different InputHandlers as for example StylusInputHandler.h
+ *
+ */
+
+#pragma once
+
 #include "control/settings/SettingsEnums.h"
 
+// forward declaration to avoid cyclic dependencies
+class ToolHandler;
+class Settings;
+
 class InputUtils {
 public:
-    static void applyButton(ToolHandler* toolHandler, Settings* settings, Buttons button, bool& pressed,
-                            bool& toolChanged, bool& configChanged);
+    /**
+     * @brief Change the active tool based on the pressed button and its corresponding settings.
+     * In case the current tool is already selected nothing is done.
+     *
+     * @param toolHandler the toolHandler containing the tools
+     * @param settings settings storing the information for the buttons
+     * @param button the button pressed
+     * @return true if the tool was changed as a result of the function.
+     * @return false if the tool was not changed as it already was pointing to the correct tool.
+     */
+    static bool applyButton(ToolHandler* toolHandler, Settings* settings, Button button);
+
+    /**
+     * @brief Warn the user in case of Questionable Settings in for the Touchscreen "Tool"
+     * Warn the user in case all of the following are true:
+     *  - Disable Drawing for this device is set for Touchscreen
+     *  - Tool - don't change is set for Touchscreen
+     *  - current Tool in the toolbar is a Drawing tool (i.e. Pen/Highlighter/Eraser)
```

Comment:
Sure that this is the combination of settings where this happens? I thought if we disable drawing for touchscreen then it should not draw at all (independent of which tool is selected). But maybe I have that wrong in my head.

Reply:
I just double checked and I basically left the logic as is (maybe apart from not returning early which only complicated the code).

This is the corresponding piece of code previous to my changes is:

```cpp
        // If an action is defined we do it, even if it's a drawing action...
        if (cfg->getDisableDrawing() && cfg->getAction() == TOOL_NONE) {
            ToolType tool = toolHandler->getToolType();
            if (tool == TOOL_PEN || tool == TOOL_ERASER || tool == TOOL_HILIGHTER) {
                g_message("ignore touchscreen for drawing!\n");
                return true;
            }
        }
```

To be honest I don't fully understand this feature anyway but at least the comment that was there initially suggests that this behaviour was intentional.

Reply:
Well that is kind of my bad as I did not comment this part good enough as it seems. What this code does is drop all inputs from the touchscreen if drawing via touchscreen is not activated. You might want to take a look into `cfg->getDisableDrawing()` as it is not named really well and might have warped logic.

Reply:
Hmm, not sure I understand. `getDisableDrawing()` seems to be quite simple actually.

```cpp
auto ButtonConfig::getDisableDrawing() const -> bool { return this->disableDrawing; }
```

What I don't understand in the above logic is: Why is it not sufficient to check only for `if (cfg->getDisableDrawing())` if the drawing should be simply disabled when that flag is set. In the current implementation (even prior to my change) setting a drawing tool in the Touch config would allow to draw again or having for example having the hand tool selected in the toolbar.

Is this really the expected bahviour? The early stopping I can easily implement.

Reply:
One cpp noob question:

I was looking throught the usages of `disableDrawing` and I was wonering why it only has a getter but no setter.

I also checked and to my understanding it is a private variable https://github.com/xournalpp/xournalpp/blob/ddbb7c6b12b4bca61be65281219767b52e67e0ae/src/control/settings/ButtonConfig.h#L41

However, there are pieces of the code (https://github.com/xournalpp/xournalpp/blob/c0855853aa49000c14c61257930cb40cc4503e34/src/control/settings/Settings.cpp#L675) where it is still set directly. I always thought that is not possible? What am I missing here?

Reply:
Look up what a friend class is ;)

Reply:
Best would be if you talk with the author of #2435 so your modifications match in this area. What I implemented back then was what the old system provided in functionality. There is definitely room for improvement.

Reply:
> Look up what a friend class is ;)

To tell the truth I do not really like friend classes. They are a good tool for unit testing as you can check for correctness even in the private state of a class but in all other cases they are quite often a sign for bad design choices. Needing access to private state of a class is just a badly designed interface.

Reply:
Hi @personalizedrefrigerator,

Looking at the files changed in your PR I don't see that we should get a conflict in functionality. But I think @LittleHuba is right, and it's better to double check.
I think the two biggest changes that are contained in my PR that could potentially lead to conflicts are:
 * the `changeTool` methods in the Input Handlers no longer return a bool
 * I refactored out the part in the `TouchDrawingInputHandler` (https://github.com/xournalpp/xournalpp/pull/2339/files#diff-c3385b4accee35899531b66ef3248e28162ee01e5d1d8ca69e5954eaedf8b13eL78)

Also if you could have a look at the comment above this comment-thread and it's implementation (https://github.com/xournalpp/xournalpp/pull/2339/files#diff-cd8df78806cf06e70a7662cb62424d7d95c728d13950575c785befaf9b345fd3R25) that would be great. Maybe you understand what the purpose of these lines was (original before refactoring: https://github.com/xournalpp/xournalpp/pull/2339/files#diff-c3385b4accee35899531b66ef3248e28162ee01e5d1d8ca69e5954eaedf8b13eL83). Otherwise I would for now try to leave the original logic untouched.

Reply:
The only change to `TouchDrawingInputHandler::changeTool` (relative to your changes) should be adding `this->updateKineticScrollingEnabled()` near line `86` (this might lead to a merge conflict, but it should be easy to resolve ☺).

Thank you for checking!


As to `if (cfg->getDisableDrawing() && cfg->getAction() == TOOL_NONE) {`, I am very confused...

`cfg` should be a `ButtonConfig`. [`getDisableDrawing`](https://github.com/xournalpp/xournalpp/blob/dbc96baec2b733b9bc050f8e6d5cc385f3a2116d/src/control/settings/ButtonConfig.cpp#L17) just returns `this->disableDrawing`, which is only set in two places:
 * In the [`ButtonConfig` constructor](https://github.com/xournalpp/xournalpp/blob/dbc96baec2b733b9bc050f8e6d5cc385f3a2116d/src/control/settings/ButtonConfig.cpp#L17) (set to `false`).
 * In [`ButtonConfigGui`](https://github.com/xournalpp/xournalpp/blob/dbc96baec2b733b9bc050f8e6d5cc385f3a2116d/src/gui/dialog/ButtonConfigGui.cpp#L239).
    * Set based on the contents of a checkbox in [`/ui/settingsButtonConfig.glade`](https://github.com/xournalpp/xournalpp/blob/dbc96baec2b733b9bc050f8e6d5cc385f3a2116d/ui/settingsButtonConfig.glade#L48).
    * `settingsButtonConfig.glade` is only referenced in `po/*.po` files and [`ButtonConfigGui.cpp`](https://github.com/xournalpp/xournalpp/blob/dbc96baec2b733b9bc050f8e6d5cc385f3a2116d/src/gui/dialog/ButtonConfigGui.cpp#L27).
        * It seems only to be constructed in [`SettingsDialog.cpp`](https://github.com/xournalpp/xournalpp/blob/dbc96baec2b733b9bc050f8e6d5cc385f3a2116d/src/gui/dialog/SettingsDialog.cpp#L146), which is shown in `Control::showSettings`, which is called by `Control::actionPerformed`.
    * The `ButtonConfigGui`s are stored in `SettingsDialog::buttonConfigs` and are only used to call `bcg->saveSettings` on `SettingsDialog::save()`.
    * `ButtonConfigGui` has a [`show` method](https://github.com/xournalpp/xournalpp/blob/dbc96baec2b733b9bc050f8e6d5cc385f3a2116d/src/gui/dialog/ButtonConfigGui.cpp#L190), with the contents `// Not implemented! This is not a dialog!`.

From the above, I am confused. Why does `settingsButtonConfig.glade` exist if `ButtonConfigGui` is never shown? How can one access the settings in `ButtonConfigGui`?

Reply:
@personalizedrefrigerator: 
I played around a bit with the debugger and I think I can explain it.

When starting the application `cfg->disableDrawing` is set in: https://github.com/xournalpp/xournalpp/blob/master/src/control/settings/Settings.cpp#L541-L548

The ButtonConfigGui is used, I think it only says `// Not implemented!` as it is not a standalone dialog but part of the Settings Dialog.
![image](https://user-images.githubusercontent.com/16611056/100526553-204cf000-31ca-11eb-8bab-4b637639bb82.png)

Within ButtonConfigGui it is then changed within https://github.com/xournalpp/xournalpp/blob/master/src/gui/dialog/ButtonConfigGui.cpp#L236-L240

So how it is set I think is explained by this.


Regarding the logic @LittleHuba I checked out the master branch before I started contributing and there the `ignore touchscreen for drawin` message was also shown even though the drawing was still happening. I then realized that this might have been a bug and I think I found the right spot where to fix it (https://github.com/xournalpp/xournalpp/pull/2339/commits/c2b94fc765672c250f8fd739496df68b628e9d94#diff-a03baeadae3022a320f793c7565ed67b183f6755149f324133430b629ee8956bR73-R74).

One code style question: For this functionality to work I had to implement all `changeTool`s to return a `bool`. Theoretically this could also be handled by throwing and catching errors, but this might be overseen by others more easily. Is the returning `bool` approach clean?

Reply:
Ahh nice that you worked all of that out so quickly. Returning a bool is preferable to throwing errors here as error catching has quite the time penalty attached to it which we would like to avoid here.